File: infra-market-server/src/main/kotlin/io/infra/market/service/AuthService.kt
package io.infra.market.service

import io.infra.market.dto.ApiResponse
import io.infra.market.dto.LoginRequest
import io.infra.market.dto.LoginResponse
import io.infra.market.dto.UserDto
import io.infra.market.dto.PermissionDto
import io.infra.market.dto.ChangePasswordRequest
import io.infra.market.enums.PermissionTypeEnum
import io.infra.market.enums.StatusEnum
import io.infra.market.repository.dao.UserDao
import io.infra.market.repository.dao.UserRoleDao
import io.infra.market.repository.dao.RolePermissionDao
import io.infra.market.repository.dao.PermissionDao
import io.infra.market.util.DateTimeUtil
import io.infra.market.util.AesUtil
import io.infra.market.util.JwtUtil
import org.springframework.stereotype.Service

@Service
class AuthService(
    private val userDao: UserDao,
    private val userRoleDao: UserRoleDao,
    private val rolePermissionDao: RolePermissionDao,
    private val permissionDao: PermissionDao,
    private val tokenService: TokenService
) {
    
    fun login(request: LoginRequest): ApiResponse<LoginResponse> {
        val user = userDao.findByUsername(request.username) ?: return ApiResponse.error("用户名或密码错误")
        
        // 验证密码
        if (!AesUtil.matches(request.password, user.password ?: "")) {
            return ApiResponse.error("用户名或密码错误")
        }
        
        // 检查用户状态
        if (user.status != StatusEnum.ACTIVE.code) {
            return ApiResponse.error("用户已被禁用")
        }
        
        // 更新登录时间
        user.lastLoginTime = System.currentTimeMillis()
        userDao.updateById(user)
        
        // 获取用户权限
        val permissions = getUserPermissions(user.id ?: 0)
        
        // 生成JWT token
        val token = JwtUtil.generateToken(user.id ?: 0, user.username ?: "")
        
        // 保存token到Redis
        tokenService.saveToken(user.id ?: 0, token)
        
        val userDto = UserDto(
            id = user.id ?: 0,
            username = user.username ?: "",
            email = user.email,
            phone = user.phone,
            status = user.status,
            lastLoginTime = DateTimeUtil.formatDateTime(user.lastLoginTime),
            createTime = DateTimeUtil.formatDateTime(user.createTime),
            updateTime = DateTimeUtil.formatDateTime(user.updateTime)
        )
        
        val response = LoginResponse(
            token = token,
            user = userDto,
            permissions = permissions
        )
        
        return ApiResponse.success(response)
    }
    
    fun logout(userId: Long): ApiResponse<Unit> {
        // 从Redis中删除token
        tokenService.deleteToken(userId)
        
        return ApiResponse.success()
    }
    
    fun getCurrentUser(userId: Long): ApiResponse<LoginResponse> {
        
        val user = userDao.findByUid(userId) ?: return ApiResponse.error("用户不存在")
        
        // 检查用户状态
        if (user.status != StatusEnum.ACTIVE.code) {
            return ApiResponse.error("用户已被禁用")
        }
        
        val permissions = getUserPermissions(user.id ?: 0)
        
        val userDto = UserDto(
            id = user.id ?: 0,
            username = user.username ?: "",
            email = user.email,
            phone = user.phone,
            status = user.status,
            lastLoginTime = DateTimeUtil.formatDateTime(user.lastLoginTime),
            createTime = DateTimeUtil.formatDateTime(user.createTime),
            updateTime = DateTimeUtil.formatDateTime(user.updateTime)
        )
        
        val response = LoginResponse(
            token = "", // 不再返回token，因为前端已经有token了
            user = userDto,
            permissions = permissions
        )
        
        return ApiResponse.success(response)
    }
    
    fun getUserMenus(userId: Long): ApiResponse<List<PermissionDto>> {
        
        val user = userDao.findByUid(userId) ?: return ApiResponse.error("用户不存在")
        
        // 检查用户状态
        if (user.status != StatusEnum.ACTIVE.code) {
            return ApiResponse.error("用户已被禁用")
        }
        
        // 获取用户角色
        val userRoles = userRoleDao.findByUserId(user.id ?: 0)
        val roleIds = userRoles.mapNotNull { it.roleId }

        if (roleIds.isEmpty()) {
            return ApiResponse.success(emptyList())
        }
        
        // 批量获取角色权限 - 避免N+1查询
        val rolePermissions = rolePermissionDao.findByRoleIds(roleIds)
        val permissionIds = rolePermissions.mapNotNull { it.permissionId }.toSet()
        
        if (permissionIds.isEmpty()) {
            return ApiResponse.success(emptyList())
        }
        
        // 获取用户拥有的所有权限（包括按钮权限）
        val allUserPermissions = permissionDao.findByIds(permissionIds.toList())
        
        // 收集所有需要的菜单权限ID
        val allMenuPermissionIds = mutableSetOf<Long>()
        
        // 收集所有需要查询的父级权限ID
        val parentIdsToQuery = mutableSetOf<Long>()
        
        // 收集所有按钮权限的父级权限ID
        for (permission in allUserPermissions) {
            if (permission.type == PermissionTypeEnum.BUTTON.code) {
                val parentId = permission.parentId
                if (parentId != null) {
                    parentIdsToQuery.add(parentId)
                }
            }
        }
        
        // 收集所有菜单权限的父级权限ID
        for (permission in allUserPermissions) {
            if (permission.type == PermissionTypeEnum.MENU.code) {
                val parentId = permission.parentId
                if (parentId != null) {
                    parentIdsToQuery.add(parentId)
                }
            }
        }
        
        // 递归收集所有父级权限ID
        val allParentIds = collectAllParentIds(parentIdsToQuery)
        
        // 批量获取所有父级权限，避免N+1查询
        val allParentPermissions = if (allParentIds.isNotEmpty()) {
            permissionDao.findByIds(allParentIds.toList())
        } else {
            emptyList()
        }
        
        // 构建权限ID到权限对象的映射
        val permissionMap = allParentPermissions.associateBy { it.id ?: 0 }
        
        // 收集所有需要的菜单权限ID（包括父级链）
        for (permission in allUserPermissions) {
            if (permission.type == PermissionTypeEnum.BUTTON.code) {
                var currentParentId = permission.parentId
                while (currentParentId != null) {
                    allMenuPermissionIds.add(currentParentId)
                    val parentPermission = permissionMap[currentParentId]
                    currentParentId = parentPermission?.parentId
                }
            }
        }
        
        for (permission in allUserPermissions) {
            if (permission.type == PermissionTypeEnum.MENU.code) {
                var currentParentId = permission.parentId
                while (currentParentId != null) {
                    allMenuPermissionIds.add(currentParentId)
                    val parentPermission = permissionMap[currentParentId]
                    currentParentId = parentPermission?.parentId
                }
            }
        }
        
        // 添加用户直接拥有的菜单权限
        allMenuPermissionIds.addAll(allUserPermissions.filter { it.type == PermissionTypeEnum.MENU.code }.map { it.id ?: 0 })
        
        // 合并所有需要的权限ID
        val allNeededPermissionIds = (allMenuPermissionIds + allUserPermissions.mapNotNull { it.id }).toSet()
        
        // 批量获取所有需要的权限详情
        val permissions = permissionDao.findByIds(allNeededPermissionIds.toList()).map { permission ->
            PermissionDto(
                id = permission.id ?: 0,
                name = permission.name ?: "",
                code = permission.code ?: "",
                type = permission.type,
                parentId = permission.parentId,
                path = permission.path,
                icon = permission.icon,
                sort = permission.sort,
                status = permission.status,
                createTime = DateTimeUtil.formatDateTime(permission.createTime),
                updateTime = DateTimeUtil.formatDateTime(permission.updateTime)
            )
        }
        
        // 构建菜单树
        val menuTree = buildMenuTree(permissions)
        
        return ApiResponse.success(menuTree)
    }
    
    private fun buildMenuTree(permissions: List<PermissionDto>): List<PermissionDto> {
        val permissionMap = permissions.associateBy { it.id }
        val rootMenus = mutableListOf<PermissionDto>()
        
        for (permission in permissions) {
            if (permission.parentId == null) {
                // 根节点，构建菜单树
                val menuWithChildren = buildMenuWithChildren(permission, permissionMap)
                rootMenus.add(menuWithChildren)
            }
        }
        
        return rootMenus.sortedBy { it.sort }
    }
    
    private fun buildMenuWithChildren(permission: PermissionDto, permissionMap: Map<Long, PermissionDto>): PermissionDto {
        val children = mutableListOf<PermissionDto>()
        
        // 查找所有以当前权限为父级的权限
        for (childPermission in permissionMap.values) {
            if (childPermission.parentId == permission.id) {
                children.add(buildMenuWithChildren(childPermission, permissionMap))
            }
        }
        
        // 按sort字段排序子权限
        val sortedChildren = children.sortedBy { it.sort }
        
        // 返回带有子权限的新PermissionDto
        return permission.copy(children = sortedChildren)
    }
    
    fun refreshToken(userId: Long): ApiResponse<Map<String, String>> {
        
        val user = userDao.findByUid(userId) ?: return ApiResponse.error("用户不存在")
        
        // 检查用户状态
        if (user.status != StatusEnum.ACTIVE.code) {
            return ApiResponse.error("用户已被禁用")
        }
        
        // 生成新的token
        val newToken = tokenService.refreshToken(userId, user.username ?: "")
        
        return ApiResponse.success(mapOf("token" to newToken))
    }
    
    fun changePassword(userId: Long, request: ChangePasswordRequest): ApiResponse<Unit> {
        
        val user = userDao.findByUid(userId) ?: return ApiResponse.error("用户不存在")
        
        // 检查用户状态
        if (user.status != StatusEnum.ACTIVE.code) {
            return ApiResponse.error("用户已被禁用")
        }
        
        // 验证旧密码
        if (!AesUtil.matches(request.oldPassword, user.password ?: "")) {
            return ApiResponse.error("原密码错误")
        }
        
        // 验证新密码不能与旧密码相同
        if (request.oldPassword == request.newPassword) {
            return ApiResponse.error("新密码不能与原密码相同")
        }
        
        // 验证新密码长度
        if (request.newPassword.length < 6) {
            return ApiResponse.error("新密码长度不能少于6位")
        }
        
        // 加密新密码
        val encryptedNewPassword = AesUtil.encrypt(request.newPassword)
        
        // 更新密码
        user.password = encryptedNewPassword
        userDao.updateById(user)
        
        return ApiResponse.success()
    }
    
    private fun getUserPermissions(userId: Long): List<String> {
        val userRoles = userRoleDao.findByUserId(userId)
        val roleIds = userRoles.mapNotNull { it.roleId }

        if (roleIds.isEmpty()) {
            return emptyList()
        }
        
        // 批量获取角色权限 - 避免N+1查询
        val rolePermissions = rolePermissionDao.findByRoleIds(roleIds)
        val permissionIds = rolePermissions.mapNotNull { it.permissionId }.toSet()
        
        if (permissionIds.isEmpty()) {
            return emptyList()
        }
        
        // 批量获取权限详情 - 避免N+1查询
        val permissions = permissionDao.findByIds(permissionIds.toList())
            .filter { it.status == StatusEnum.ACTIVE.code }
            .map { it.code ?: "" }
        
        return permissions.distinct()
    }
    
    /**
     * 递归收集所有父级权限ID
     */
    private fun collectAllParentIds(initialParentIds: Set<Long>): Set<Long> {
        val allParentIds = mutableSetOf<Long>()
        val processedIds = mutableSetOf<Long>()
        val idsToProcess = initialParentIds.toMutableSet()
        
        while (idsToProcess.isNotEmpty()) {
            val currentBatch = idsToProcess.toList()
            idsToProcess.clear()
            
            // 批量查询当前批次的权限
            val permissions = permissionDao.findByIds(currentBatch)
            
            for (permission in permissions) {
                val permissionId = permission.id ?: 0
                if (permissionId !in processedIds) {
                    processedIds.add(permissionId)
                    allParentIds.add(permissionId)
                    
                    // 如果这个权限还有父级，添加到下一批处理
                    val parentId = permission.parentId
                    if (parentId != null && parentId !in processedIds) {
                        idsToProcess.add(parentId)
                    }
                }
            }
        }
        
        return allParentIds
    }
    

}

File: infra-market-server-go/internal/service/auth_service.go
package service

import (
	"log"
	"time"

	"github.com/bucketheadv/infra-market/internal/dto"
	"github.com/bucketheadv/infra-market/internal/entity"
	"github.com/bucketheadv/infra-market/internal/repository"
	"github.com/bucketheadv/infra-market/internal/util"
)

type AuthService struct {
	userRepo           *repository.UserRepository
	userRoleRepo       *repository.UserRoleRepository
	rolePermissionRepo *repository.RolePermissionRepository
	permissionRepo     *repository.PermissionRepository
	tokenService       *TokenService
}

func NewAuthService(
	userRepo *repository.UserRepository,
	userRoleRepo *repository.UserRoleRepository,
	rolePermissionRepo *repository.RolePermissionRepository,
	permissionRepo *repository.PermissionRepository,
	tokenService *TokenService,
) *AuthService {
	return &AuthService{
		userRepo:           userRepo,
		userRoleRepo:       userRoleRepo,
		rolePermissionRepo: rolePermissionRepo,
		permissionRepo:     permissionRepo,
		tokenService:       tokenService,
	}
}

// Login 用户登录
func (s *AuthService) Login(req dto.LoginRequest) dto.ApiData[dto.LoginResponse] {
	user, err := s.userRepo.FindByUsername(req.Username)
	if err != nil {
		return dto.Error[dto.LoginResponse]("用户名或密码错误", 400)
	}

	// 验证密码
	if !util.Matches(req.Password, user.Password) {
		return dto.Error[dto.LoginResponse]("用户名或密码错误", 400)
	}

	// 检查用户状态
	if user.Status != "active" {
		return dto.Error[dto.LoginResponse]("用户已被禁用", 400)
	}

	// 更新登录时间
	now := time.Now().UnixMilli()
	user.LastLoginTime = &now
	_ = s.userRepo.Update(user)

	// 获取用户权限
	permissions := s.getUserPermissions(user.ID)

	// 生成JWT token
	token, err := util.GenerateToken(user.ID, user.Username)
	if err != nil {
		log.Printf("用户登录失败，生成token失败，用户名: %s, 错误: %v\n", req.Username, err)
		return dto.Error[dto.LoginResponse]("生成token失败", 500)
	}

	// 保存token到Redis
	s.tokenService.SaveToken(user.ID, token)

	userDto := s.convertUserToDto(user, nil)
	response := dto.LoginResponse{
		Token:       token,
		User:        userDto,
		Permissions: permissions,
	}

	return dto.Success(response)
}

// Logout 用户登出
func (s *AuthService) Logout(uid uint64) dto.ApiData[any] {
	s.tokenService.DeleteToken(uid)
	return dto.Success[any](nil)
}

// GetCurrentUser 获取当前用户信息
func (s *AuthService) GetCurrentUser(uid uint64) dto.ApiData[dto.LoginResponse] {
	user, err := s.userRepo.FindByUID(uid)
	if err != nil {
		log.Printf("获取当前用户信息失败，用户ID: %d, 错误: %v\n", uid, err)
		return dto.Error[dto.LoginResponse]("用户不存在", 404)
	}

	if user.Status != "active" {
		return dto.Error[dto.LoginResponse]("用户已被禁用", 400)
	}

	permissions := s.getUserPermissions(user.ID)
	userDto := s.convertUserToDto(user, nil)

	response := dto.LoginResponse{
		Token:       "", // 不再返回token
		User:        userDto,
		Permissions: permissions,
	}

	return dto.Success(response)
}

// GetUserMenus 获取用户菜单
func (s *AuthService) GetUserMenus(uid uint64) dto.ApiData[[]dto.PermissionDto] {
	user, err := s.userRepo.FindByUID(uid)
	if err != nil {
		log.Printf("获取用户菜单失败，用户ID: %d, 错误: %v\n", uid, err)
		return dto.Error[[]dto.PermissionDto]("用户不存在", 404)
	}

	if user.Status != "active" {
		return dto.Error[[]dto.PermissionDto]("用户已被禁用", 400)
	}

	// 获取用户的所有权限（包括按钮权限）
	allUserPermissions := s.getUserPermissionList(uid)

	if len(allUserPermissions) == 0 {
		return dto.Success([]dto.PermissionDto{})
	}

	// 收集所有需要查询的父级权限ID
	parentIdsToQuery := make(map[uint64]bool)

	// 收集所有按钮权限和菜单权限的父级权限ID
	for _, permission := range allUserPermissions {
		if permission.ParentID != nil {
			parentIdsToQuery[*permission.ParentID] = true
		}
	}

	// 递归收集所有父级权限ID
	allParentIds := s.collectAllParentIds(parentIdsToQuery)

	// 批量获取所有父级权限，避免N+1查询
	var allParentPermissions []entity.Permission
	if len(allParentIds) > 0 {
		parentIdsSlice := make([]uint64, 0, len(allParentIds))
		for id := range allParentIds {
			parentIdsSlice = append(parentIdsSlice, id)
		}
		allParentPermissions, _ = s.permissionRepo.FindByIDs(parentIdsSlice)
	}

	// 构建权限ID到权限对象的映射
	permissionMap := make(map[uint64]entity.Permission)
	for _, p := range allParentPermissions {
		permissionMap[p.ID] = p
	}

	// 收集所有需要的菜单权限ID（包括父级链）
	allMenuPermissionIds := make(map[uint64]bool)

	// 从按钮权限向上收集父级菜单
	for _, permission := range allUserPermissions {
		if permission.Type == "button" && permission.ParentID != nil {
			currentParentID := permission.ParentID
			for currentParentID != nil {
				allMenuPermissionIds[*currentParentID] = true
				// 从 permissionMap 中查找父级权限
				if parent, ok := permissionMap[*currentParentID]; ok && parent.ParentID != nil {
					currentParentID = parent.ParentID
				} else {
					// 如果不在 permissionMap 中，说明已经到根节点或权限不存在
					break
				}
			}
		}
	}

	// 从菜单权限向上收集父级菜单
	for _, permission := range allUserPermissions {
		if permission.Type == "menu" {
			// 向上收集父级菜单
			if permission.ParentID != nil {
				currentParentID := permission.ParentID
				for currentParentID != nil {
					allMenuPermissionIds[*currentParentID] = true
					// 从 permissionMap 中查找父级权限
					if parent, ok := permissionMap[*currentParentID]; ok && parent.ParentID != nil {
						currentParentID = parent.ParentID
					} else {
						// 如果不在 permissionMap 中，说明已经到根节点或权限不存在
						break
					}
				}
			}
		}
	}

	// 合并所有需要的权限ID（包括所有父级权限和用户直接拥有的权限）
	for id := range allParentIds {
		allMenuPermissionIds[id] = true
	}
	for _, permission := range allUserPermissions {
		allMenuPermissionIds[permission.ID] = true
	}

	// 批量获取所有需要的权限详情
	allNeededPermissionIds := make([]uint64, 0, len(allMenuPermissionIds))
	for id := range allMenuPermissionIds {
		allNeededPermissionIds = append(allNeededPermissionIds, id)
	}
	allPermissions, _ := s.permissionRepo.FindByIDs(allNeededPermissionIds)

	// 只保留菜单类型的权限
	menuPermissions := make([]entity.Permission, 0)
	for _, p := range allPermissions {
		if p.Type == "menu" {
			menuPermissions = append(menuPermissions, p)
		}
	}

	// 构建权限树
	tree := s.buildPermissionTree(menuPermissions)

	return dto.Success(tree)
}

// RefreshToken 刷新token
func (s *AuthService) RefreshToken(uid uint64) dto.ApiData[map[string]string] {
	user, err := s.userRepo.FindByUID(uid)
	if err != nil {
		return dto.Error[map[string]string]("用户不存在", 404)
	}

	token, err := util.GenerateToken(user.ID, user.Username)
	if err != nil {
		log.Printf("刷新token失败，生成token失败，用户ID: %d, 错误: %v\n", uid, err)
		return dto.Error[map[string]string]("生成token失败", 500)
	}

	s.tokenService.SaveToken(uid, token)

	return dto.Success(map[string]string{"token": token})
}

// ChangePassword 修改密码
func (s *AuthService) ChangePassword(uid uint64, req dto.ChangePasswordRequest) dto.ApiData[any] {
	user, err := s.userRepo.FindByUID(uid)
	if err != nil {
		return dto.Error[any]("用户不存在", 404)
	}

	// 验证原密码
	if !util.Matches(req.OldPassword, user.Password) {
		return dto.Error[any]("原密码错误", 400)
	}

	// 加密新密码
	encryptedPassword, err := util.Encrypt(req.NewPassword)
	if err != nil {
		log.Printf("修改密码失败，密码加密失败，用户ID: %d, 错误: %v\n", uid, err)
		return dto.Error[any]("密码加密失败", 500)
	}

	user.Password = encryptedPassword
	if err := s.userRepo.Update(user); err != nil {
		log.Printf("修改密码失败，更新密码失败，用户ID: %d, 错误: %v\n", uid, err)
		return dto.Error[any]("更新密码失败", 500)
	}

	return dto.Success[any](nil)
}

// getUserPermissions 获取用户权限编码列表
func (s *AuthService) getUserPermissions(uid uint64) []string {
	userRoles, err := s.userRoleRepo.FindByUID(uid)
	if err != nil {
		return []string{}
	}
	if len(userRoles) == 0 {
		return []string{}
	}

	roleIDs := make([]uint64, 0)
	for _, ur := range userRoles {
		if ur.RoleID != nil {
			roleIDs = append(roleIDs, *ur.RoleID)
		}
	}

	if len(roleIDs) == 0 {
		return []string{}
	}

	rolePermissions, err := s.rolePermissionRepo.FindByRoleIDs(roleIDs)
	if err != nil {
		return []string{}
	}
	if len(rolePermissions) == 0 {
		return []string{}
	}

	// 收集权限ID并去重
	permissionIDSet := make(map[uint64]bool)
	for _, rp := range rolePermissions {
		permissionIDSet[rp.PermissionID] = true
	}

	if len(permissionIDSet) == 0 {
		return []string{}
	}

	// 转换为切片
	permissionIDs := make([]uint64, 0, len(permissionIDSet))
	for id := range permissionIDSet {
		permissionIDs = append(permissionIDs, id)
	}

	permissions, err := s.permissionRepo.FindByIDs(permissionIDs)
	if err != nil {
		return []string{}
	}

	// 过滤激活状态的权限并去重
	codeSet := make(map[string]bool)
	codes := make([]string, 0)
	for _, p := range permissions {
		// 只返回激活状态的权限
		if p.Status == "active" && p.Code != "" {
			// 去重
			if !codeSet[p.Code] {
				codeSet[p.Code] = true
				codes = append(codes, p.Code)
			}
		}
	}

	return codes
}

// getUserPermissionList 获取用户权限实体列表
func (s *AuthService) getUserPermissionList(uid uint64) []entity.Permission {
	userRoles, _ := s.userRoleRepo.FindByUID(uid)
	if len(userRoles) == 0 {
		return []entity.Permission{}
	}

	roleIDs := make([]uint64, 0)
	for _, ur := range userRoles {
		if ur.RoleID != nil {
			roleIDs = append(roleIDs, *ur.RoleID)
		}
	}

	rolePermissions, _ := s.rolePermissionRepo.FindByRoleIDs(roleIDs)
	if len(rolePermissions) == 0 {
		return []entity.Permission{}
	}

	permissionIDs := make([]uint64, 0)
	for _, rp := range rolePermissions {
		permissionIDs = append(permissionIDs, rp.PermissionID)
	}

	permissions, _ := s.permissionRepo.FindByIDs(permissionIDs)
	return permissions
}

// collectAllParentIds 递归收集所有父级权限ID
func (s *AuthService) collectAllParentIds(parentIds map[uint64]bool) map[uint64]bool {
	allParentIds := make(map[uint64]bool)
	for id := range parentIds {
		allParentIds[id] = true
	}

	// 查询这些父级权限
	if len(parentIds) == 0 {
		return allParentIds
	}

	parentIdsSlice := make([]uint64, 0, len(parentIds))
	for id := range parentIds {
		parentIdsSlice = append(parentIdsSlice, id)
	}

	parentPermissions, _ := s.permissionRepo.FindByIDs(parentIdsSlice)

	// 收集这些父级权限的父级ID
	nextLevelParentIds := make(map[uint64]bool)
	for _, p := range parentPermissions {
		if p.ParentID != nil && !allParentIds[*p.ParentID] {
			nextLevelParentIds[*p.ParentID] = true
		}
	}

	// 如果有新的父级ID，递归收集
	if len(nextLevelParentIds) > 0 {
		recursiveParents := s.collectAllParentIds(nextLevelParentIds)
		for id := range recursiveParents {
			allParentIds[id] = true
		}
	}

	return allParentIds
}

// buildPermissionTree 构建权限树
func (s *AuthService) buildPermissionTree(permissions []entity.Permission) []dto.PermissionDto {
	// 先转换为DTO
	permissionMap := make(map[uint64]dto.PermissionDto)
	for _, p := range permissions {
		permissionMap[p.ID] = s.convertPermissionToDto(p, nil)
	}

	return buildPermissionTreeFromMap(permissions, permissionMap)
}

// convertUserToDto 转换用户实体为DTO
func (s *AuthService) convertUserToDto(user *entity.User, roleIds []uint64) dto.UserDto {
	var lastLoginTime *string
	if user.LastLoginTime != nil {
		formatted := util.Format(user.LastLoginTime)
		lastLoginTime = &formatted
	}

	return dto.UserDto{
		ID:            user.ID,
		Username:      user.Username,
		Email:         user.Email,
		Phone:         user.Phone,
		Status:        user.Status,
		LastLoginTime: lastLoginTime,
		RoleIds:       roleIds,
		CreateTime:    util.Format(&user.CreateTime),
		UpdateTime:    util.Format(&user.UpdateTime),
	}
}

// convertPermissionToDto 转换权限实体为DTO
func (s *AuthService) convertPermissionToDto(permission entity.Permission, children []dto.PermissionDto) dto.PermissionDto {
	return dto.PermissionDto{
		ID:         permission.ID,
		Name:       permission.Name,
		Code:       permission.Code,
		Type:       permission.Type,
		ParentID:   permission.ParentID,
		Path:       permission.Path,
		Icon:       permission.Icon,
		Sort:       permission.Sort,
		Status:     permission.Status,
		Children:   children,
		CreateTime: util.Format(&permission.CreateTime),
		UpdateTime: util.Format(&permission.UpdateTime),
	}
}
